:construction: WIP on diff

Andreas-Espelund opened the pull request:
# Diff av manifester

Diff kommandoen lar deg se hvordan json output endrer seg basert på endringer du har gjort.
Kommandoen tar inn en commit hash som flagg (default er HEAD) som bestemmer hvilken 
versjon av den samme filen du skal sammenligne med.


## Eksempel på bruk

```jsonnet

// slik ser filen ut ved forrige commit 

{
  person: {
    name: 'Alice',
    age: 30 + 12 + 'bob',
    hobbies: [
      'reading',
      'cycling',
    ] + ['reading'],
  },
}

// nå ser filen slik ut

{
  person: {
    name: 'Alice',
    age: 30 + 12 + 'bob',
    hobbies: [
      'reading',
      'cycling',
    ] + ['reading'],
  },
} + {
  person+: {
    hobbies+: [
      'swimming',
    ],
  },
}

```

```shell
# ved å kjøre diff kommandoen
skipctl  manifests diff --path=testdata/jsonnet/diff/another_diff.jsonnet


# får man dette som output
time=2025-09-15T17:37:27.609+02:00 level=INFO msg=diff file=testdata/jsonnet/diff/another_diff.jsonnet commit_hash=HEAD
-         "reading"
+         "reading",
+         "swimming"
```

hauks1 commented:
Ser det kanskje litt snodig ut med vanlig tekst? 🤔 Hvis det er lange string id'er så blir det kanskje litt komplisert å se hva som faktisk er diffen?

<img width="1329" height="423" alt="Screenshot 2025-09-16 at 08 36 07" src="https://github.com/user-attachments/assets/182d50ca-2419-48b8-994a-243e6b4b742f" />

evenh commented on pkg/utils/files.go:
Dette burde høre hjemme i `pkg/diff` eller et mer passende sted. Farger er fint, men hvordan skal man bruke dette videre i automatiseringsformål som GitHub Actions/Workflows?

evenh commented on pkg/manifest/document.go:
Alt dette hører ikke hjemme i `Document` structen imho

evenh commented on pkg/manifest/differ.go:
```diff
@@ -0,0 +1,66 @@
+	if err != nil {
+		return err
+	}
+	// BELOW IS THE NORMAL THAT IS LOCATED IN document.go
+	// prevFile, err := file.FromPrevHash(d.prevHash)
+	// if err != nil {
+	// 	return err
+	// }
```
?

evenh commented on pkg/manifest/document.go:
```diff
@@ -20,15 +23,124 @@ type Document struct {
-	r.Content = content
-	if r.FromStdin {
-		if _, err := io.WriteString(os.Stdout, r.Content); err != nil {
+func (d *Document) FromPrevHashWithGoGit(hash string) (*Document, error) {
```
go-git er bare en implementasjonsdetalj og bør ikke ligge i navnet

Andreas-Espelund commented:
**Changes** 
- Added support for diffing against branch names, and renamed --hash flag to --ref.
- Changed regex to support branch names
- Moved util funcitons into `/utils/diff.go` file.
- Improved call to exec

okpedersen reviewed (commented):
Mye bra her! 🚀 Noe småpirk, og noen litt større greier. I tillegg til de andre kommentarene lurer jeg på om det hadde vært lurt med flere tester 🤔 Kjenner ikke testriggen så godt i dette repoet, er endringer i testdata nok eller burde det vært mer?

evenh commented on pkg/manifest/git_importer.go:
```diff
@@ -0,0 +1,27 @@
+	return &GitImporter{Ref: ref}
+}
+
+func (gi *GitImporter) Import(importedFrom, importedPath string) (jsonnet.Contents, string, error) {
```
Ah, da forstår jeg. Da gir det mening 💡

evenh commented on pkg/diff/diff_test.go:
Tenker bruk av [testify](https://github.com/stretchr/testify) hadde gjort seg her. Må ikke som en del av denne PR'en

Andreas-Espelund replied:
Det noteres!
